### PR TITLE
API Gateway monitoring [OPS-1575]

### PR DESCRIPTION
**Pull request by bilalahmad99**

### CODE REVIEW

**Comment by bilalahmad99** [comment]
@e0d

**Comment by e0d** [comment]
@bilalahmad99 there are a bunch of linting issues in the PR, you can see them here:
https://travis-ci.org/edx/api-manager/builds/135902552

You can also run ```make quality``` locally to see them.

File: scripts/aws/lambda_exec_policy.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "logs:CreateLogGroup",
+        "logs:CreateLogStream",
+        "logs:PutLogEvents"
+      ],
+      "Resource": "arn:aws:logs:*:*:*"
```
Comment (e0d) [inline scripts/aws/lambda_exec_policy.json]: We should make this more targeted.

File: scripts/aws/monitor.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        EvaluationPeriods=eval_period,
+        Threshold=threshold,
+        ComparisonOperator=comparison,
+        AlarmActions=[topic])
+
+    logging.info('response for creating api alarm = "%s"',
+                 response['ResponseMetadata'])
+
+
+def create_firehose_stream(firehose, stream_name, arn, bucket_arn):
```
Comment (e0d) [inline scripts/aws/monitor.py]: This is no longer needed.

File: scripts/aws/monitor.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        })
+
+    logging.info('response for creating firehose stream = "%s"',
+                 response)
+
+
+def create_lambda_function_zip(splunk_host, splunk_token):
+    """Updates and Zips the lambda function file"""
+
+    with open('index.js', "wt") as lambda_function_file:
```
Comment (e0d) [inline scripts/aws/monitor.py]: Let's use jinja here to make this more maintainable.

File: scripts/aws/monitor.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        with open('index.js.template', 'rt') as lambda_function_template:
+            for line in lambda_function_template:
+                lambda_function_file.write(line.replace('<splunk host:port>',
+                                                        splunk_host).replace('<token>',
```
Comment (e0d) [inline scripts/aws/monitor.py]: Let's also write the altered file to the system defined temp directory.  This will allow the OS to deal with cleanup and not dirty the git workspace.

File: scripts/aws/monitor.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        lambda_zip.write('index.js')
+
+    return zip_file
+
+
+def role_exists(iam, role_name):
+    """Checks if the role exists already"""
+    try:
+        iam.get_role(RoleName=role_name)
+    except botocore.exceptions.ClientError:
```
Comment (e0d) [inline scripts/aws/monitor.py]: Is there a more specific exception that could be checked here?

File: scripts/aws/monitor.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    response = client.list_topics()
+    for value in response['Topics']:
+        sns_arn = value['TopicArn']
+        if topic_name in str(sns_arn):
+            return str(sns_arn)
+        else:
+            return None
+
+
+def cleanup():
```
Comment (e0d) [inline scripts/aws/monitor.py]: As discussed, look at temp dir context manager.

File: scripts/aws/monitor.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                     4, 300, 1, [{'Name': 'ApiName', 'Value': args.api_gw}, {'Name': 'Label', 'Value': args.api_gw}],
+                     get_topic_arn(sns_client, 'aws-non-critical-alert'))
+
+    iam_client = session.create_client('iam', args.aws_region)
+    role_arn = create_role(iam_client, 'lambda_basic_execution_monitor',
+                           '{"Version": "2012-10-17","Statement": [{"Effect": "Allow","Principal": '
+                           '{"Service": "lambda.amazonaws.com"},"Action": "sts:AssumeRole"}]}',
+                           open(os.path.join(os.path.dirname(__file__), 'lambda_exec_policy.json')).read())
+
+    time.sleep(10)
```
Comment (e0d) [inline scripts/aws/monitor.py]: Why is this necessary?  Minimally, add a comment.

File: scripts/aws/monitor.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    lambda_client = session.create_client('lambda', args.aws_region)
+    zip_file_name = create_lambda_function_zip(args.splunk_host, args.splunk_token)
+    create_lambda_function(lambda_client, 'cloudwatch-logs-splunk', 'nodejs4.3', role_arn,
+                           'index.handler', zip_file_name,
+                           'Demonstrates logging events to Splunk HTTP Event '
+                           'Collector.', 10, 512)
+    time.sleep(2)
```
Comment (e0d) [inline scripts/aws/monitor.py]: Why sleep here?

File: scripts/aws/index.js.template [hunk 9]
```diff
@@ -0,0 +1,187 @@
+/**
```
Comment (e0d) [inline scripts/aws/index.js.template]: Currently reaching out to confirm what license the blueprint is under.

File: scripts/aws/monitor.py (hunk 6 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: done

File: scripts/aws/monitor.py (hunk 7 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: done

File: scripts/aws/monitor.py (hunk 5 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: Apparently not. It seems to be the only one.
https://github.com/boto/boto3/issues/167

File: scripts/aws/monitor.py (hunk 4 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: done

File: scripts/aws/monitor.py (hunk 3 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: done

File: scripts/aws/monitor.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+def create_lambda_function(client, function_name, runtime, role,
+                           handler, zip_file, description, timeout, mem_size):
+    """Create a lambda function to pull data from cloudwatch event"""
+    response = client.create_function(
+        FunctionName=function_name,
+        Runtime=runtime,
+        Role=role,
+        Handler=handler,
+        Code={
+            'ZipFile': open(zip_file, 'rb').read(),
```
Comment (e0d) [inline scripts/aws/monitor.py]: This file is never closed.

File: scripts/aws/monitor.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+    j2_env = Environment(loader=FileSystemLoader(os.path.join(os.path.dirname(__file__), 'templates')),
+                         trim_blocks=False)
+    splunk_values = {
+        'splunk_ip': splunk_host,
+        'token': splunk_token,
+    }
+    js_file = temp_dir + '/index.js'
+    with open(js_file, 'w') as lambda_function_file:
+        lf_data = j2_env.get_template('index.j2').render(splunk_values)
```
Comment (e0d) [inline scripts/aws/monitor.py]: small think, but let's make the template index.js.j2, this way it's clear this is a jinja template of a js file.

File: scripts/aws/monitor.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    logging.info('Waiting for the newly created role to be available')
+    # Sleep for 10 seconds to allow the role created above to be avialable for lambda function creation
+    time.sleep(10)
+    lambda_client = session.create_client('lambda', args.aws_region)
+    tmpdirname = tempfile.mkdtemp()
+    zip_file_name = create_lambda_function_zip(tmpdirname, args.splunk_host, args.splunk_token)
+    create_lambda_function(lambda_client, 'cloudwatch-logs-splunk', 'nodejs4.3', role_arn,
+                           'index.handler', zip_file_name,
+                           'Demonstrates logging events to Splunk HTTP Event '
+                           'Collector.', 10, 512)
```
Comment (efagin) [inline scripts/aws/monitor.py]: What variables are 10 and 512 and why are they fixed values instead of configurable tuning parameters? (Based on my knowledge of Lambda I assume that this is "max lambda execution time" and "max lambda memory utilization".)

File: scripts/aws/monitor.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+                     4, 300, 1, [{'Name': 'ApiName', 'Value': args.api_gw}, {'Name': 'Label', 'Value': args.api_gw}],
+                     get_topic_arn(sns_client, 'aws-non-critical-alert'))
+
+    create_api_alarm(cw, 'api_errors_5xx', '5XXError', 'ApiGateway', 'Average',
+                     'GreaterThanOrEqualToThreshold', 'Average 5XX errors for a period of 5 min',
+                     4, 300, 1, [{'Name': 'ApiName', 'Value': args.api_gw}, {'Name': 'Label', 'Value': args.api_gw}],
+                     get_topic_arn(sns_client, 'aws-non-critical-alert'))
+
+    iam_client = session.create_client('iam', args.aws_region)
+    role_arn = create_role(iam_client, 'lambda_basic_execution_monitor',
```
Comment (efagin) [inline scripts/aws/monitor.py]: What happens if this role already exists? Also, consider choosing a more detailed name for this role, as "monitor" is not very specific.

File: scripts/aws/monitor.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+                           '{"Service": "lambda.amazonaws.com"},"Action": "sts:AssumeRole"}]}',
+                           open(os.path.join(os.path.dirname(__file__), 'lambda_exec_policy.json')).read())
+
+    logging.info('Waiting for the newly created role to be available')
+    # Sleep for 10 seconds to allow the role created above to be avialable for lambda function creation
+    time.sleep(10)
+    lambda_client = session.create_client('lambda', args.aws_region)
+    tmpdirname = tempfile.mkdtemp()
+    zip_file_name = create_lambda_function_zip(tmpdirname, args.splunk_host, args.splunk_token)
+    create_lambda_function(lambda_client, 'cloudwatch-logs-splunk', 'nodejs4.3', role_arn,
```
Comment (efagin) [inline scripts/aws/monitor.py]: Similar question here - what happens if this lambda function already exists? (How is the lambda code going to be updated? By re-running this script or something else?)

File: scripts/aws/lambda_exec_policy.json (hunk 1 above)
Comment (efagin) [inline scripts/aws/lambda_exec_policy.json]: Consider making this policy a Jinja template as well, as the exact resource should be different for different environments. I don't think we'd ever want to allow a Lambda to put log events in any CloudWatch Logs stream it wants.

File: scripts/aws/monitor.py (hunk 12 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: Made them configurable

File: scripts/aws/monitor.py (hunk 13 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: The function checks if the role already exists. Also changed the name of the role to bit more detailed. See if its okay.

File: scripts/aws/monitor.py (hunk 14 above)
Comment (bilalahmad99) [inline scripts/aws/monitor.py]: If the function already exists it updates the code of the function, otherwise creates a new function.

File: scripts/aws/lambda_exec_policy.json (hunk 1 above)
Comment (bilalahmad99) [inline scripts/aws/lambda_exec_policy.json]: Done

**Comment by efagin** [comment]
👍 - just merged to master. thanks!
